components/ciphertool/src/main/java/org/wso2/ciphertool/utils/KeyStoreUtil.java

```diff
@@ -46,59 +52,123 @@ public static Cipher initializeCipher() {
         return cipher;
     }
 
-    private static KeyStore getKeyStore(String location, String storePassword, String storeType) {
-        BufferedInputStream bufferedInputStream = null;
-        try {
-            bufferedInputStream = new BufferedInputStream(new FileInputStream(location));
-            KeyStore keyStore = KeyStore.getInstance(storeType);
+    private static KeyStore getKeyStore(String location, String storePassword, String storeType, String providerName) {
+        KeyStore keyStore;
+        try (BufferedInputStream bufferedInputStream = new BufferedInputStream(Files.
+                newInputStream(Paths.get(location)))) {
+            String provider = getPreferredJceProvider(providerName);
+            if (provider != null) {
+                keyStore = KeyStore.getInstance(storeType, provider);
+            } else {
+                keyStore = KeyStore.getInstance(storeType);
+            }
             keyStore.load(bufferedInputStream, storePassword.toCharArray());
             return keyStore;
         } catch (KeyStoreException e) {
-            throw new CipherToolException("Error loading keyStore from ' " + location + " ' ", e);
+            throw new CipherToolException(
+                    "KeyStore type is not supported or not initialized properly while loading from '" +
+                            location + "'", e);
         } catch (IOException e) {
-            throw new CipherToolException("Error loading keyStore from ' " + location + " ' ", e);
+            throw new CipherToolException(
+                    "I/O error occurred while reading the KeyStore from '" + location + "'", e);
         } catch (NoSuchAlgorithmException e) {
-            throw new CipherToolException("Error loading keyStore from ' " + location + " ' ", e);
+            throw new CipherToolException(
+                    "Integrity check algorithm for the KeyStore is not available while loading from '" +
+                            location + "'", e);
         } catch (CertificateException e) {
-            throw new CipherToolException("Error loading keyStore from ' " + location + " ' ", e);
-        } finally {
-            if (bufferedInputStream != null) {
-                try {
-                    bufferedInputStream.close();
-                } catch (IOException e) {
-                    System.err.println("Error while closing input stream");
-                }
+            throw new CipherToolException(
+                    "One or more certificates in the KeyStore could not be loaded or parsed from '" +
+                            location + "'", e);
+        } catch (NoSuchProviderException e) {
+            throw new CipherToolException("Specified security provider is not available in this environment", e);
+        }
+    }
+
+    public static void addJceProvider(String providerName) {
+        String jceProvider = getPreferredJceProvider(providerName);

```

Comment:
We are returning null here for things other than BC provider and BC FIPS provider, need to check what happens for other providers.

Reply:
Execute with the default provider. https://github.com/wso2/cipher-tool/pull/95/files#diff-649dd1a961290405d3a1eafc89e1b576872e14c02884e70a4a3992f908c07fe2R119